common/src/main/java/com/microsoft/identity/common/internal/platform/DevicePopManager.java

```diff
@@ -0,0 +1,834 @@
+        }
+
+        final KeyPairGeneratorSpec spec = specBuilder.build();
+        keyPairGenerator.initialize(spec);
+    }
+
+    /**
+     * Gets the current time as a {@link Date}.
+     *
+     * @param calendar The {@link Calendar} implementation to use.
+     * @return The current time.
+     */
+    private static Date getNow(@NonNull final Calendar calendar) {
+        return calendar.getTime();
+    }
+
+    /**
+     * For the supplied {@link KeyStore.Entry}, get a corresponding {@link KeyPair} instance.
+     *
+     * @param entry The Keystore.Entry to use.
+     * @return The resulting KeyPair.
+     */
+    private static KeyPair getKeyPairForEntry(@NonNull final KeyStore.Entry entry) {
+        final PrivateKey privateKey = ((KeyStore.PrivateKeyEntry) entry).getPrivateKey();
+        final PublicKey publicKey = ((KeyStore.PrivateKeyEntry) entry).getCertificate().getPublicKey();
+        return new KeyPair(publicKey, privateKey);
+    }
+
+    /**
+     * Gets the corresponding {@link RSAKey} for the supplied {@link KeyPair}.
+     *
+     * @param keyPair The KeyPair to use.
+     * @return The resulting RSAKey.
+     */
+    private static RSAKey getRsaKeyForKeyPair(@NonNull final KeyPair keyPair) {
+        return new RSAKey.Builder((RSAPublicKey) keyPair.getPublic())
+                .privateKey(keyPair.getPrivate())
+                .keyUse(null)
+                .build();
+    }
+
+    /**
+     * Gets the base64url encoded public jwk for the supplied RSAKey.
+     *
+     * @param rsaKey The input key material.
+     * @return The base64url encoded jwk.
+     */
+    private static String getReqCnfForRsaKey(@NonNull final RSAKey rsaKey)
+            throws JOSEException, JSONException {
+        final String thumbprintStr = getThumbprintForRsaKey(rsaKey);
+        final String thumbprintMinifiedJson =
+                new JSONObject()
+                        .put("kid", thumbprintStr)
+                        .toString();
+
+        return base64UrlEncode(thumbprintMinifiedJson);
+    }
+
+    private static String getThumbprintForRsaKey(@NonNull RSAKey rsaKey) throws JOSEException {
+        final Base64URL thumbprint = rsaKey.computeThumbprint();
+        return thumbprint.toString();
+    }
+
+    /**
+     * Encodes a String with Base64Url and no padding.
+     *
+     * @param input String to be encoded.
+     * @return Encoded result from input.
+     */
+    private static String base64UrlEncode(@NonNull final String input) {
+        String result = null;
+
+        try {
+            byte[] encodeBytes = input.getBytes(ENCODING_SCHEME);
+            result = Base64.encodeToString(
+                    encodeBytes,
+                    Base64.NO_PADDING | Base64.NO_WRAP | Base64.URL_SAFE
+            );
+        } catch (final UnsupportedEncodingException e) {
+            e.printStackTrace();
+        }
+
+        return result;
+    }
+
+    /**
+     * Returns the cnf claim used in SHRs; format is JSON.
```

Comment:
nit: what does SHR stand for ?

Reply:
"Signed HTTP Request" -- I'll put this in the comments

Reply:
Clarified in ed4be06a411ff0543985b5da19672407d04f994e